src/throttler/throttler.ts

```diff
@@ -91,20 +96,39 @@ export abstract class Throttler<T> {
     return p;
   }
 
-  public add(task: T): void {
+  /**
+   * Add the task to the throttler.
+   * When the task is completed, resolve will be called with handler's result.
+   * If this task fails after retries, reject will be called with the error.
+   */
+  public add(
+    task: T,
+    wait?: { resolve: (result: any) => void; reject: (err: Error) => void }
+  ): void {
     if (this.closed) {
       throw new Error("Cannot add a task to a closed throttler.");
     }
-
     if (!this.startTime) {
       this.startTime = Date.now();
     }
-
-    this.tasks[this.total] = task;
+    this.taskDatas[this.total] = {
+      task,
+      wait,
+      retryCount: 0,
+    };
     this.total++;
     this.process();
   }
 
+  /**
+   * Add the task to the throttler and return a promise of handler's result.
+   */
+  public throttle<R>(task: T): Promise<R> {
```

Comment:
I am open for a better way of doing this.
Right now, the caller will define a type, and it will cast it.

The result type may vary for a single `Throttler` instance, but is the same for each specific call of `throttle`.